.ci/magician/cmd/generate_comment.go

```diff
@@ -0,0 +1,379 @@
+		if err != nil {
+			fmt.Println("Diff processor error: ", err)
+			diffProccessorErr = err
+		}
+	}
+	for _, path := range []string{"old", "new", "bin"} {
+		if err := r.RemoveAll(filepath.Join(diffProcessorPath, path)); err != nil {
+			return "", err
+		}
+	}
+	for _, path := range []string{"old", "new"} {
+		if err := r.Copy(tpgbLocalPath, filepath.Join(diffProcessorPath, path)); err != nil {
+			return "", err
+		}
+	}
+
+	if diffProccessorErr != nil {
+		fmt.Println("Breaking changes failed")
+		breakingChanges = `## Breaking Change Detection Failed
+The breaking change detector crashed during execution. This is usually due to the downstream provider(s) failing to compile. Please investigate or follow up with your reviewer.`
+	} else {
+		fmt.Println("Breaking changes succeeded")
+		breakingChanges = compareBreakingChanges(tpgBreaking, tpgbBreaking)
+	}
+	return breakingChanges, nil
+}
+
+// Get the breaking change message including the unique tpg messages and all tpgb messages.
+func compareBreakingChanges(tpgBreaking, tpgbBreaking string) string {
+	var allMessages []string
+	if tpgBreaking == "" {
+		if tpgbBreaking == "" {
+			return ""
+		}
+		allMessages = strings.Split(tpgbBreaking, "\n")
+	} else if tpgbBreaking == "" {
+		allMessages = strings.Split(tpgBreaking, "\n")
+	} else {
+		dashExp := regexp.MustCompile("-.*")
+		tpgMessages := strings.Split(tpgBreaking, "\n")
+		tpgbMessages := strings.Split(tpgbBreaking, "\n")
+		tpgbSet := make(map[string]struct{}, len(tpgbMessages))
+		var tpgUnique []string
+		for _, message := range tpgbMessages {
+			simple := dashExp.ReplaceAllString(message, "")
+			tpgbSet[simple] = struct{}{}
+		}
+		for _, message := range tpgMessages {
+			simple := dashExp.ReplaceAllString(message, "")
+			if _, ok := tpgbSet[simple]; !ok {
+				tpgUnique = append(tpgUnique, message)
+			}
+		}
+		allMessages = append(tpgUnique, tpgbMessages...)
+	}
+	if len(allMessages) > 0 {
+		return `Breaking Change(s) Detected
+The following breaking change(s) were detected within your pull request.
+
+* ` + strings.Join(allMessages, "\n* ") + `
+
+If you believe this detection to be incorrect please raise the concern with your reviewer.
+If you intend to make this change you will need to wait for a [major release](https://www.terraform.io/plugin/sdkv2/best-practices/versioning#example-major-number-increments) window.
+An ` + "`override-breaking-change`" + `label can be added to allow merging.
+`
+	}
+	return ""
+}
+
+// Run the missing test detector and return the results.
+// Returns an empty string unless there are missing tests.
+// Error will be nil unless an error occurs during setup.
+func detectMissingTests(mmLocalPath, tpgbLocalPath, oldBranch string, r gcRunner) (string, error) {
+	tpgbLocalPathOld := tpgbLocalPath + "old"
+	if err := r.Copy(tpgbLocalPath, tpgbLocalPathOld); err != nil {
+		return "", err
+	}
+	r.Chdir(tpgbLocalPathOld)
```

Comment:
This might be a bit confusing because a function is mutating the runners directory. So the passed in runner has a different directory when popping up. Can we manage this slightly better?

Reply:
Maybe by saving the old directory and switching back to it before returning to the parent function ?

Reply:
oldDirectory := r.Getcwd()
defer r.Chdir(oldDirectory)